### PR TITLE
Bug 1813233: DEVEXP-533 - ocm k8s 118; prometheus bump; *-go bump

### PR DESCRIPTION
**Pull request by gabemontero**
@openshift/openshift-team-developer-experience 

might need an openshift/release golang 1.13 bump (that config file is somewhat atypical)

### CODE REVIEW

**Comment by gabemontero** [comment]
> @gabemontero I'm pretty sure k8s 1.18 requires golang 1.13. Worth checking that we are using the right golang version in openshift/release.

yeah per my note in the description @adambkaplan it is and it isn't .... tracking down the genesis of https://github.com/openshift/release/blob/master/ci-operator/config/openshift/openshift-controller-manager/openshift-openshift-controller-manager-master.yaml#L14-L22

**Comment by gabemontero** [comment]
Turns out that extra stanza was intended and is pattern master team has employed in a few repos for speeding up builds.

https://github.com/openshift/release/pull/8104 is up

**Comment by gabemontero** [comment]
Per https://github.com/openshift/release/pull/8104#issuecomment-608054118 the rehearsal job's image build failed in the same way even after changing that second golang reference to 1.13

Going to create a debug commit here to dive into things a bit

**Comment by gabemontero** [comment]
ok the images job completes now @adambkaplan  ... it was the yum install within the golang 1.13 build 

I'll work it in #4-dev-triage and #forum-testplatform, but I'm pretty sure those 2 devel packages are nice to have vs. necessities, if push comes to shove

**Comment by gabemontero** [comment]
but since the use of the golang-1.13 image worked for me locally, it somehow feels like a CI RPM mirror issue

**Comment by gabemontero** [comment]
So yeah those RPMs are already installed in the 1.12 image ... see https://github.com/openshift/openshift-controller-manager/pull/86#issuecomment-608108920

they are not in the 1.13 image, and the repo/mirror configurations may be different

I've engaged with @jupierce

**Comment by gabemontero** [comment]
github outage for the e2e-aws-builds failure ... going to try switching RUN yum install and WORKDIR to rule that out

**Comment by gabemontero** [comment]
And it worked with the yum install before WORKDIR .... maybe the docker/imagebuilder versions in those two golang images images varies enough that `WORKDIR` responds differently.

But in any event, those two packages in in 1.13 as well:

```
STEP 1: FROM registry.svc.ci.openshift.org/ocp/builder:golang-1.13 AS builder
Getting image source signatures
Copying blob sha256:5c983eb30936c925f524779b7cc5fa14bb5988607a43ecf44cfeae996899d0f9
Copying blob sha256:ab5ef0e5819490abe86106fd9f4381123e37a03e80e650be39f7938d30ecb530
Copying config sha256:581802e18145a03794653def8d64fdc36fba124b05eec2d89588b9292276c810
Writing manifest to image destination
Storing signatures
STEP 2: ENV "foo"="bar"
STEP 3: RUN yum install -y gpgme-devel libassuan-devel
Loaded plugins: fastestmirror, ovl
Determining fastest mirrors
 * base: d36uatko69830t.cloudfront.net
 * epel: d2lzkl7pfhq30w.cloudfront.net
 * extras: d36uatko69830t.cloudfront.net
 * updates: d36uatko69830t.cloudfront.net
http://csc.mcs.sdsmt.edu/epel/7/x86_64/repodata/c9cce1169888adb7cda0c77b82fdd5ffff0f81029bd5048dfde95824f3adfb33-updateinfo.xml.bz2: [Errno 12] Timeout on http://csc.mcs.sdsmt.edu/epel/7/x86_64/repodata/c9cce1169888adb7cda0c77b82fdd5ffff0f81029bd5048dfde95824f3adfb33-updateinfo.xml.bz2: (28, 'Operation too slow. Less than 1000 bytes/sec transferred the last 30 seconds')
Trying other mirror.
Package gpgme-devel-1.3.2-5.el7.x86_64 already installed and latest version
Package libassuan-devel-2.1.0-3.el7.x86_64 already installed and latest version
Nothing to do
``

so all we are doing is asking for yum flakes in the future trying this unnecessary install @adambkaplan 

I'm removing it officially.

**Comment by gabemontero** [comment]
ok we've got all green tests for the OCM k8s rebase @adambkaplan

**Comment by ricardomaraschini** [comment]
@gabemontero I have reasons to believe that this PR is also going to solve a BZ[1] I am working on. It seems like now we listen on all protocol versions instead of only on tcp4[2].

Could we have this backported to 4.4 ?

[1] https://bugzilla.redhat.com/show_bug.cgi?id=1813233
[2] https://github.com/openshift/library-go/commit/886b6c5cb61ef66f9f0f5c934022366746cffef8#diff-eb5800da2d439f81e8e41059183544b6

**Comment by gabemontero** [comment]
> @gabemontero I have reasons to believe that this PR is also going to solve a BZ[1] I am working on. It seems like now we listen on all protocol versions instead of only on tcp4[2].
> 
> Could we have this backported to 4.4 ?
> 
> ~[1] https://bugzilla.redhat.com/show_bug.cgi?id=1813233~
> [1] https://bugzilla.redhat.com/show_bug.cgi?id=1816948
> [2] [openshift/library-go@886b6c5#diff-eb5800da2d439f81e8e41059183544b6](https://github.com/openshift/library-go/commit/886b6c5cb61ef66f9f0f5c934022366746cffef8#diff-eb5800da2d439f81e8e41059183544b6)

@ricardomaraschini - I just did a `git diff-tree --no-commit-id --name-only -r` on both of the commits in this PR and neither of them list ` pkg/config/configdefaults/config_default.go `, which is the file that comes up when I click on your [2] link.

Doing a `git blame` on config_default.go in this repo's vendor tree, it was @adambkaplan 's https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/137 that pulled in that change.

So switch gears and ask in that PR.

**Comment by ricardomaraschini** [comment]
> > @gabemontero I have reasons to believe that this PR is also going to solve a BZ[1] I am working on. It seems like now we listen on all protocol versions instead of only on tcp4[2].
> > Could we have this backported to 4.4 ?
> > ~[1] https://bugzilla.redhat.com/show_bug.cgi?id=1813233~
> > [1] https://bugzilla.redhat.com/show_bug.cgi?id=1816948
> > [2] [openshift/library-go@886b6c5#diff-eb5800da2d439f81e8e41059183544b6](https://github.com/openshift/library-go/commit/886b6c5cb61ef66f9f0f5c934022366746cffef8#diff-eb5800da2d439f81e8e41059183544b6)
> 
> @ricardomaraschini - I just did a `git diff-tree --no-commit-id --name-only -r` on both of the commits in this PR and neither of them list `pkg/config/configdefaults/config_default.go`, which is the file that comes up when I click on your [2] link.
> 
> Doing a `git blame` on config_default.go in this repo's vendor tree, it was @adambkaplan 's [openshift/cluster-openshift-controller-manager-operator#137](https://github.com/openshift/cluster-openshift-controller-manager-operator/pull/137) that pulled in that change.
> 
> So switch gears and ask in that PR.

This seems to be introduced by this PR:

https://github.com/openshift/openshift-controller-manager/pull/85/files#diff-bc17ec995aceeacd885829ee8e6a2d6a

**Comment by gabemontero** [comment]
My bad ... too man rebase PRs ... my https://github.com/openshift/openshift-controller-manager/pull/85#issuecomment-609792727 comment was working off of ocm-o, not ocm

But as @ricardomaraschini and I discussed on slack, pulling in a k8s rebase into 4.4 may be too onerous.

He is first going to try and craft a 4.4 rebase PR to just that library-go commit level, and see if it avoids bumping k8s

**Comment by ricardomaraschini** [comment]
/retitle Bug 1813233: DEVEXP-533 - ocm k8s 118; prometheus bump; *-go bump

File: Dockerfile [hunk 1]
```diff
@@ -1,6 +1,5 @@
 FROM registry.svc.ci.openshift.org/ocp/builder:golang-1.13 AS builder
 WORKDIR /go/src/github.com/openshift/openshift-controller-manager
-RUN yum install -y gpgme-devel libassuan-devel
```
Comment (ricardomaraschini) [inline Dockerfile:3]: Any special reason why was this here? I mean, I am having problems [building release 4.4](https://github.com/openshift/openshift-controller-manager/pull/88) of this repository, most likely due to these dependencies. I may go ahead and remove them as well but I would like to check with you first.

Comment (gabemontero) [reply Dockerfile:3]: Goes back to the origins of this stuff (even predates back to when all of this was in origin).

As time has passed, these installs were done in the base image, and as I was told the clean up here had simply been overlooked.

From my investigations `registry.svc.ci.openshift.org/ocp/builder:golang-1.12` had these RPMs as well.  The yum install simply was less flaky as a result of an imagebuilder diff between the 1.12 and 1.13 images.

Bottom line, I would +1 @ricardomaraschini removing that line in the 4.4 dockerfiles.

File: pkg/apps/deploymentconfig/controller_ref_manager.go [hunk 2]
```diff
@@ -34,7 +35,7 @@ var _ RCControlInterface = &RealRCControl{}
 
 // PatchReplicationController executes a strategic merge patch contained in 'data' on RC specified by 'namespace' and 'name'
 func (r RealRCControl) PatchReplicationController(namespace, name string, data []byte) error {
-	_, err := r.KubeClient.CoreV1().ReplicationControllers(namespace).Patch(name, types.StrategicMergePatchType, data)
+	_, err := r.KubeClient.CoreV1().ReplicationControllers(namespace).Patch(context.TODO(), name, types.StrategicMergePatchType, data, kmetav1.PatchOptions{})
```
Comment (adambkaplan) [inline pkg/apps/deploymentconfig/controller_ref_manager.go:38]: /cc @tnozicka

File: pkg/apps/metrics/prometheus/metrics.go [hunk 3]
```diff
@@ -96,6 +96,16 @@ type failedRollout struct {
...
 }
 
+func (c *appsCollector) ClearState() {
+	c.createLock.Lock()
+	defer c.createLock.Unlock()
+	c.isCreated = false
+}
+
+func (c *appsCollector) FQName() string {
+	return "openshift_apps_deploymentconfigs"
```
Comment (adambkaplan) [inline pkg/apps/metrics/prometheus/metrics.go:106]: /cc @tnozicka

File: pkg/authorization/defaultrolebindings/defaultrolebindings.go [hunk 4]
```diff
@@ -124,7 +126,7 @@ func (c *DefaultRoleBindingController) syncNamespace(namespaceName string) error
 			continue
 		}
 
-		_, err := c.roleBindingClient.RoleBindings(namespaceName).Create(&desiredRoleBinding)
+		_, err := c.roleBindingClient.RoleBindings(namespaceName).Create(context.TODO(), &desiredRoleBinding, metav1.CreateOptions{})
```
Comment (adambkaplan) [inline pkg/authorization/defaultrolebindings/defaultrolebindings.go:129]: /cc @deads2k

**Comment by gabemontero** [comment]
bump @adambkaplan

File: pkg/route/ingress/ingress.go [hunk 5]
```diff
@@ -434,15 +435,15 @@ func (c *Controller) sync(key queueKey) error {
 			return err
 		}
 		data = []byte(fmt.Sprintf(`[{"op":"replace","path":"/spec","value":%s}]`, data))
-		_, err = c.client.Routes(route.Namespace).Patch(route.Name, types.JSONPatchType, data)
+		_, err = c.client.Routes(route.Namespace).Patch(context.TODO(), route.Name, types.JSONPatchType, data, metav1.PatchOptions{})
```
Comment (adambkaplan) [inline pkg/route/ingress/ingress.go:438]: /cc @ironcladlou

File: dependencymagnet/doc.go [hunk 6]
```diff
@@ -1,7 +1,9 @@
...
 
 import (
-	// alpha-build-machinery has non-go code needed for makefiles.
-	// Needed to work with go modules.
-	_ "github.com/openshift/library-go/alpha-build-machinery"
+	_ "github.com/openshift/build-machinery-go/make"
+	_ "github.com/openshift/build-machinery-go/make/lib"
+	_ "github.com/openshift/build-machinery-go/make/targets"
+	_ "github.com/openshift/build-machinery-go/make/targets/golang"
+	_ "github.com/openshift/build-machinery-go/make/targets/openshift"
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply dependencymagnet/doc.go:8]: if you are more precise and list multiple things though you do not pull in unneeded items 

IMO this is better

File: pkg/apps/deployer/deployer_controller.go [hunk 7]
```diff
@@ -160,7 +161,7 @@ func (c *DeploymentController) handle(deployment *corev1.ReplicationController,
 				return fatalError(fmt.Sprintf("couldn't make deployer pod for %q: %v", appsutil.LabelForDeployment(deployment), err))
 			}
 			// Create the deployer pod.
-			deploymentPod, err := c.pn.Pods(deployment.Namespace).Create(deployerPod)
+			deploymentPod, err := c.pn.Pods(deployment.Namespace).Create(context.TODO(), deployerPod, metav1.CreateOptions{})
```
[the remark this answers is not in the document]
Comment (gabemontero) [reply pkg/apps/deployer/deployer_controller.go:164]: Not sure what you mean by "separate" context @adambkaplan  ... this is the first call to create a context I see in that method.

And of course you have to pass in a context.

Can you clarify?

[the remark this answers is not in the document]
Comment (adambkaplan) [reply pkg/apps/deployer/deployer_controller.go:164]: @gabemontero `context` is something typically defined in a main method and passed down  - in our case through controllers. Doing so gives us nice things like timeouts and graceful termination.

We can keep this as TODO for now and address later as tech debt.

File: dependencymagnet/doc.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (adambkaplan) [reply dependencymagnet/doc.go:8]: My worry is that this is brittle - `build-machinery-go` is a bag of Makefiles that may be moved about. If a dependency is added to a folder outside of `lib` or `targets`, we won't pick it up.

[the remark this answers is not in the document]
Comment (adambkaplan) [reply dependencymagnet/doc.go:8]: Per discussion - I've been convinced otherwise to be more precise. Approving.

File: pkg/apps/deployer/deployer_controller.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (tnozicka) [reply pkg/apps/deployer/deployer_controller.go:164]: I am fine with `context.TODO()` to limit the scope of the rebase

File: Dockerfile [hunk 8]
```diff
@@ -1,4 +1,4 @@
-FROM registry.svc.ci.openshift.org/ocp/builder:golang-1.12 AS builder
+FROM registry.svc.ci.openshift.org/ocp/builder:golang-1.13 AS builder
```
Comment (tnozicka) [inline Dockerfile:1]: I don't think changes not matching `go.{mod,sum} vendor/*` belong to a bump commit

File: dependencymagnet/doc.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (tnozicka) [reply dependencymagnet/doc.go:8]: What reasons were that? I think it's better to claim a dependency on the whole lib (as the dependency systems are unrelated and don't know about each other) and avoid less skilled people having to search and update this magnets.

File: go.mod [hunk 9]
```diff
@@ -1,76 +1,68 @@
...
-	k8s.io/kube-proxy => k8s.io/kube-proxy v0.0.0-20191122223145-16f2c0c680a0
-	k8s.io/kube-scheduler => k8s.io/kube-scheduler v0.0.0-20191122223507-ac656df73673
-	k8s.io/kubectl => k8s.io/kubectl v0.0.0-20191122225023-1e3c8b70f494
-	k8s.io/kubelet => k8s.io/kubelet v0.0.0-20191122223325-9316382755ad
-	k8s.io/kubernetes => k8s.io/kubernetes v1.17.0-rc.1
-	k8s.io/legacy-cloud-providers => k8s.io/legacy-cloud-providers v0.0.0-20191122224431-860df69ff5cc
-	k8s.io/metrics => k8s.io/metrics v0.0.0-20191122222628-19ed227de2b6
-	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.0.0-20191122221846-294c70c3d5d4
+	// for whatever weird reason go mod would not let me set openshift/api to this level in the require section
+	github.com/openshift/api => github.com/openshift/api v0.0.0-20200326160804-ecb9283fe820
```
Comment (tnozicka) [inline go.mod:45]: Is this because your dependency vendors the wrong api level?
https://github.com/openshift/runtime-utils/blob/9169de69ebf6/go.mod#L10
I think this is worthy of fixing the dependency to avoid mistakes in the chain.

File: go.mod [hunk 10]
```diff
@@ -1,76 +1,68 @@
...
+    // these are needed since k8s.io/kubernetes cites v0.0.0 for each of these k8s deps in its go.mod
```
Comment (tnozicka) [inline go.mod:46]: could this be more limited just to what's being used in required section?

File: go.mod (hunk 9 above)
Comment (gabemontero) [reply go.mod:45]: https://github.com/openshift/runtime-utils/pull/10

File: go.mod (hunk 10 above)
Comment (gabemontero) [reply go.mod:46]: Nope.  I had to add these one by one, otherwise I got errors like

```
go: k8s.io/kubernetes@v1.18.0 requires
	k8s.io/apiextensions-apiserver@v0.0.0: reading k8s.io/apiextensions-apiserver/go.mod at revision v0.0.0: unknown revision v0.0.0
```

Which is what I was referring to in the comment above.

Even when I tried moving each of these to the require section, and remove them from the replace section, I got the same errors, like 

```
go: k8s.io/kubernetes@v1.18.0 requires
	k8s.io/api@v0.0.0: reading k8s.io/api/go.mod at revision v0.0.0: unknown revision v0.0.0
```

File: Dockerfile (hunk 8 above)
Comment (gabemontero) [reply Dockerfile:1]: will be fixed in next push

File: go.mod (hunk 10 above)
Comment (tnozicka) [reply go.mod:46]: ok, thanks for the detail

File: go.mod (hunk 9 above)
Comment (gabemontero) [reply go.mod:45]: yep making that update did the trick - thanks

will be part of next push

File: dependencymagnet/doc.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (gabemontero) [reply dependencymagnet/doc.go:8]: I went back and looked at https://github.com/openshift/openshift-apiserver and I now believe the "how much is imported" savings I thought we were going to get does not materialize.

I will reverse stance and switch to one import ref.

As an aside, it was my experience as a new consumer of build-machinery-go, I had to search through these to understand what was there and available anyway.

**Comment by gabemontero** [comment]
ok @tnozicka @adambkaplan I believe all requested changes have been pushed - thanks

**Comment by gabemontero** [comment]
aws / terraform flake e2e-aws-builds

/test e2e-aws-builds

**Comment by gabemontero** [comment]
all recent failures so far are around lower level component daemonsets not reaching 3 replicas (so *possibly* aws/env related)

will have to engage tomorrow AM if still persisting and see if I can tell if it is *NOT* aws/env

**Comment by gabemontero** [comment]
ok a passing builds and upgrade e2e

/retest

**Comment by adambkaplan** [comment]
Story to address the use of `context`: https://issues.redhat.com/browse/DEVEXP-578

**Comment by tnozicka** [comment]
everything except the bump commit having non-vendored changes lgtm, I'll leave it up with you if you fix that

**Comment by gabemontero** [comment]
ci flake `no 'cli' image was tagged into the stable-initial stream, that image is required for building a release`

**Comment by gabemontero** [comment]
more env looking install flakes

/retest

**Comment by adambkaplan** [comment]
/hold

`e2e-aws` is currently failing across the board due to install errors. Holding so the bot doesn't make the CI situation worse.

**Comment by gabemontero** [comment]
> /hold
> 
> `e2e-aws` is currently failing across the board due to install errors. Holding so the bot doesn't make the CI situation worse.

either @adambkaplan or I will unhold once CI looks a bit healthier

**Comment by gabemontero** [comment]
the latest e2e-aws at least go installed and had passing tests when I looked a few mintutes ago

**Comment by gabemontero** [comment]
and we have a passing e2e-aws https://prow.svc.ci.openshift.org/view/gcs/origin-ci-test/pr-logs/pull/openshift_openshift-controller-manager/85/pull-ci-openshift-openshift-controller-manager-master-e2e-aws/258 !!!
